src/main/drivers/serial.c

```diff
@@ -44,16 +44,16 @@ void serialWrite(serialPort_t *instance, uint8_t ch)
 }
 
 
-void serialWriteBuf(serialPort_t *instance, const uint8_t *data, int count)
+void serialWriteBufNoFlush(serialPort_t *instance, const uint8_t *data, int count)
 {
     if (instance->vTable->writeBuf) {
         instance->vTable->writeBuf(instance, data, count);
     } else {
-        for (const uint8_t *p = data; count > 0; count--, p++) {
-
-            while (!serialTxBytesFree(instance)) {
-            };
+        // The transmit buffer is large enough to hold any single message, so only wait once
```

Comment:
I've checked and the transmit buffer never gets close to filling up, so this is fallback behaviour. Rather than dropping characters this would indeed block and the task would be late and if bad enough detected by the CPU load check. I consider this preferable to random MSP dropped packets and resulting strange undefined OSD behaviour.

Reply:
And it's worth noting that the check is now only being done once which is more efficient.

Reply:
Buffer filling is bad - it will cause busy waiting, which will degrade performance. But if some code tries to write more than transmit buffer size, it will immediately crash FC.